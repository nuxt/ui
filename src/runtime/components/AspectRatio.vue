<script lang="ts">
import { tv } from 'tailwind-variants'
import type { AspectRatioProps as AspectRatioRootProps } from 'radix-vue'
import type { AppConfig } from '@nuxt/schema'
import _appConfig from '#build/app.config'
import theme from '#build/ui/aspect-ratio'

const appConfig = _appConfig as AppConfig & { ui: { aspectRatio: Partial<typeof theme> } }

const aspectRatio = tv({ extend: tv(theme), ...(appConfig.ui?.aspectRatio || {}) })

export interface AspectRatioProps extends AspectRatioRootProps {
  /**
   * The element or component this component should render as.
   * @defaultValue 'span'
   */
  as?: any
  class?: any
  ui?: Partial<typeof aspectRatio.slots>
}

export interface AspectRatioSlots {
  default(props?: {}): any
}
</script>

<script setup lang="ts">
import { AspectRatio, useForwardProps } from 'radix-vue'
import { reactivePick } from '@vueuse/core'

const props = defineProps<AspectRatioProps>()
defineSlots<AspectRatioSlots>()

const rootProps = useForwardProps(reactivePick(props, 'as', 'asChild', 'ratio'))

// eslint-disable-next-line vue/no-dupe-keys
const ui = aspectRatio()
</script>

<template>
  <AspectRatio v-bind="rootProps" :class="ui.root({ class: [props.class, props.ui?.root] })">
    <slot />
  </AspectRatio>
</template>
