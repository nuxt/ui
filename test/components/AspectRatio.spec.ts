import { describe, it, expect } from 'vitest'
import AspectRatio, { type AspectRatioProps, type AspectRatioSlots } from '../../src/runtime/components/AspectRatio.vue'
import ComponentRender from '../component-render'

describe('AspectRatio', () => {
  it.each([
    // Props
    ['with as', { props: { as: 'div' } }],
    ['with class', { props: { class: 'block' } }],
    ['with ui', { props: { ui: { root: 'block' } } }],
    ['with aspect 4/3', { props: { ratio: 4 / 3 } }],
    ['with aspect 16/9', { props: { ratio: 16 / 9 } }],
    // Slots
    ['with default slot', { slots: { default: () => 'default slot' } }]
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: AspectRatioProps, slots?: Partial<AspectRatioSlots> }) => {
    const html = await ComponentRender(nameOrHtml, options, AspectRatio)
    expect(html).toMatchSnapshot()
  })
})
