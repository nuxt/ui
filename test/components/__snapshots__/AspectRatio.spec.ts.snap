// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AspectRatio > renders with as correctly 1`] = `
"<div style="position: relative; width: 100%; padding-bottom: 100%;" data-radix-aspect-ratio-wrapper="">
  <div style="position: absolute;"></div>
</div>"
`;

exports[`AspectRatio > renders with aspect 4/3 correctly 1`] = `
"<div style="position: relative; width: 100%; padding-bottom: 75%;" data-radix-aspect-ratio-wrapper="">
  <div style="position: absolute;"></div>
</div>"
`;

exports[`AspectRatio > renders with aspect 16/9 correctly 1`] = `
"<div style="position: relative; width: 100%; padding-bottom: 56.25%;" data-radix-aspect-ratio-wrapper="">
  <div style="position: absolute;"></div>
</div>"
`;

exports[`AspectRatio > renders with class correctly 1`] = `
"<div style="position: relative; width: 100%; padding-bottom: 100%;" data-radix-aspect-ratio-wrapper="">
  <div style="position: absolute;" class="block"></div>
</div>"
`;

exports[`AspectRatio > renders with default slot correctly 1`] = `
"<div style="position: relative; width: 100%; padding-bottom: 100%;" data-radix-aspect-ratio-wrapper="">
  <div style="position: absolute;">default slot</div>
</div>"
`;

exports[`AspectRatio > renders with ui correctly 1`] = `
"<div style="position: relative; width: 100%; padding-bottom: 100%;" data-radix-aspect-ratio-wrapper="">
  <div style="position: absolute;" class="block"></div>
</div>"
`;
